api/src/prover/hypernova/error.rs

```diff
@@ -0,0 +1,145 @@
+
+/// Errors related to proof generation
+#[derive(Debug)]
+pub enum ProofError {
+    /// An error occured loading or executing program
+    NexusVMError(NexusVMError),
+
+    /// An error occurred reading file system
+    IOError(std::io::Error),
+
+    /// An error occured during circuit synthesis
+    CircuitError(SynthesisError),
+
+    /// An error occured serializing to disk
+    SerError(SerializationError),
+
+    /// The witness does not satisfy the constraints as R1CS
+    R1CSWitnessError(R1CSError),
+
+    /// The witness does not satisfy the constraints as CCS
+    CCSWitnessError(CCSError),
+
+    /// Invalid folding step index
+    InvalidIndex(usize),
+
+    /// Public Parameters do not match circuit
+    InvalidPP,
+
+    /// An error occured while computing the HyperNova folding
+    FoldingError(HNFoldingError),
+
+    /// An error occured while setting up a polynomial commitment
+    PolyCommitmentError,
+
+    /// The HyperNova prover produced an invalid proof
+    HyperNovaProofError,
+
+    /// A proof has been read from a file that does not match the expected format
+    InvalidProofFormat,
+}
+use ProofError::*;
+
+impl From<NexusVMError> for ProofError {
+    fn from(x: NexusVMError) -> ProofError {
+        NexusVMError(x)
+    }
+}
+
+impl From<std::io::Error> for ProofError {
+    fn from(x: std::io::Error) -> ProofError {
+        IOError(x)
+    }
+}
+
+impl From<SynthesisError> for ProofError {
+    fn from(x: SynthesisError) -> ProofError {
+        CircuitError(x)
+    }
+}
+
+impl From<HyperNovaError> for ProofError {
+    fn from(x: HyperNovaError) -> ProofError {
+        match x {
+            HyperNovaError::R1CS(e) => R1CSWitnessError(e),
+            HyperNovaError::CCS(e) => CCSWitnessError(e),
+            HyperNovaError::Synthesis(e) => CircuitError(e),
+            HyperNovaError::HNFolding(e) => FoldingError(e),
+            HyperNovaError::InvalidPublicInput => HyperNovaProofError,
+            HyperNovaError::PolyCommitmentSetup => PolyCommitmentError,
+        }
+    }
+}
+
+impl From<SerializationError> for ProofError {
+    fn from(x: SerializationError) -> ProofError {
+        SerError(x)
+    }
+}
+
+impl From<NovaProofError> for ProofError {
+    fn from(x: NovaProofError) -> ProofError {
+        match x {
+            NovaProofError::NexusVMError(e) => NexusVMError(e),
+            NovaProofError::IOError(e) => IOError(e),
+            NovaProofError::CircuitError(e) => CircuitError(e),
+            NovaProofError::SerError(e) => SerError(e),
+            // The above error conversions allow reusing convienence functions
+            // from the nova implemementation in this crate.
+            //
+            // The remaining errors are thrown by functions that are specific to
+            // a proof system, and so shouldn't be shared across the definitions
+            // even with identical names.
+            _ => unimplemented!(),
```

Comment:
Yeah that was me reusing folding error everywhere in the crate.

I think soon we will need to split nova into smaller crates to allow easy wasm compilation, and this is when we'll improve error handling. Can keep this for the time being.